libs/frontend/feature-build/src/lib/code-editor-plugins.ts

```diff
@@ -0,0 +1,58 @@
+import { schemaCode } from '@pubstudio/frontend/util-edit-text'
+import { Keys } from '@pubstudio/frontend/util-key-listener'
+import hljs from 'highlight.js/lib/core'
+import { keymap } from 'prosemirror-keymap'
+import { Fragment, Slice } from 'prosemirror-model'
+import { Plugin } from 'prosemirror-state'
+import { highlightPlugin } from './prosemirror-highlighthjs-plugin'
+
+// Manually create a new slice instead of using the third parameter `slice` from
+// `handlePaste` because new line characters will be ignored for unknown reason
+// when using it to do content replacement in the editor state.
+const transformPasteContentPlugin = new Plugin({
+  props: {
+    handlePaste: (view, e) => {
+      const pastedContent = e.clipboardData?.getData('text') ?? ''
+      if (pastedContent) {
+        const textNode = schemaCode.text(pastedContent)
+        const slice = new Slice(Fragment.from(textNode), 0, 0)
+        view.dispatch(view.state.tr.replaceSelection(slice))
+      }
+      return true
+    },
+  },
+})
+
+// Manually insert a new line instead of using the default behavior to avoid creating
+// a new block of <p> as a sibling of <pre>.
+// Use handleDOMEvents insead of keymap because combinations like 'Shift-Enter' didn't
+// work for some reason; our callback function never gets executed.
+const interceptNewLineEventPlugin = new Plugin({
```

Comment:
I noticed that pressing Cmd + Enter, Ctrl + Enter, or Shift + Enter will create an "incorrect" new line in the editor. I have to add this event handler plugin so that those key combinations will lead to a normal new line just like what Enter does.

https://github.com/samatechtw/pubstudio-builder/assets/59676941/692b79ea-50f8-454d-ae37-670c6ee5cc90

I use `handleDOMEvents` here insead of keymap because for some reason, combinations like 'Shift-Enter' didn't
work in keymap 😕. The callback function gets called.

Reply:
Interesting! I'm not sure why the keymap doesn't work.